interop-cats/shared/src/main/scala/zio/interop/cats.scala

```diff
@@ -35,348 +39,453 @@ abstract class CatsEffectPlatform
-    semigroupKLossyInstance0.asInstanceOf[SemigroupK[ZIO[R, E, *]]]
-
-  private[this] final val monadErrorInstance0: MonadError[ZIO[Any, Any, *], Any] =
-    new CatsMonadError[Any, Any]
+  protected[this] final val arrowChoiceInstance0: ArrowChoice[RIO] =
+    new ZioArrowChoice
 
-  private[this] final val semigroupKLossyInstance0: SemigroupK[ZIO[Any, Any, *]] =
-    new CatsSemigroupKLossy[Any, Any]
+  private[this] val monadErrorInstance0: MonadError[Task, Throwable] =
+    new ZioMonadError[Any, Throwable]
 }
 
-private class CatsDefer[R, E] extends Defer[ZIO[R, E, *]] {
-  def defer[A](fa: => ZIO[R, E, A]): ZIO[R, E, A] = ZIO.effectSuspendTotal(fa)
+private class ZioDefer[R, E] extends Defer[ZIO[R, E, *]] {
+  type F[A] = ZIO[R, E, A]
+
+  override final def defer[A](fa: => F[A]): F[A] =
+    ZIO.effectSuspendTotal(fa)
 }
 
-private class CatsConcurrentEffect[R](rts: Runtime[R])
-    extends CatsConcurrent[R]
-    with effect.ConcurrentEffect[RIO[R, *]] {
-
-  override final def runAsync[A](fa: RIO[R, A])(
-    cb: Either[Throwable, A] => effect.IO[Unit]
-  ): effect.SyncIO[Unit] =
-    effect.SyncIO {
-      rts.unsafeRunAsync(fa.run) { exit =>
-        cb(exit.flatMap(identity).toEither).unsafeRunAsync(_ => ())
-      }
-    }
+private class ZioConcurrent[R, E] extends ZioMonadError[R, E] with GenConcurrent[ZIO[R, E, *], E] {
 
-  override final def runCancelable[A](fa: RIO[R, A])(
-    cb: Either[Throwable, A] => effect.IO[Unit]
-  ): effect.SyncIO[effect.CancelToken[RIO[R, *]]] =
-    effect.SyncIO {
-      rts.unsafeRun {
-        ZIO.descriptor
-          .bracketExit(
-            (descriptor, exit: Exit[Throwable, A]) =>
-              ZIO.effectTotal {
-                exit match {
-                  case Exit.Failure(cause) if !cause.interruptors.forall(_ == descriptor.id) => ()
-                  case _ =>
-                    effect.IO.suspend(cb(exit.toEither)).unsafeRunAsync(_ => ())
-                }
-              },
-            _ => fa
-          )
-          .interruptible
-          .forkDaemon
-          .map(_.interrupt.unit)
-      }
-    }
+  private def toPoll(restore: ZIO.InterruptStatusRestore) = new Poll[ZIO[R, E, *]] {
+    override def apply[T](fa: ZIO[R, E, T]): ZIO[R, E, T] = restore(fa)
+  }
+
+  private def toFiber[A](fiber: Fiber[E, A]) = new effect.Fiber[F, E, A] {
+    override final val cancel: F[Unit]           = fiber.interrupt.unit
+    override final val join: F[Outcome[F, E, A]] = fiber.await.map(toOutcome)
+  }
+
+  override def ref[A](a: A): F[effect.Ref[F, A]] =
+    ZRef.make(a).map(new ZioRef(_))
+
+  override def deferred[A]: F[Deferred[F, A]] =
+    Promise.make[E, A].map(new ZioDeferred(_))
+
+  override final def start[A](fa: F[A]): F[effect.Fiber[F, E, A]] =
+    fa.interruptible.forkDaemon.map(toFiber)
+
+  override def never[A]: F[A] =
+    ZIO.never
+
+  override final def cede: F[Unit] =
+    ZIO.yieldNow
+
+  override final def forceR[A, B](fa: F[A])(fb: F[B]): F[B] =
```

Comment:
I think if you change the implementation to recover from all errors exception interruption (including recovering from `Die` failures) then we can `Die` if a finalizer fails in `onCancel`, `guarantee`, or `bracket`.

Reply:
Interesting - do you have something like this in mind? 
```scala
    fa.run.flatMap {
      case Exit.Success(_)                          => fb
      case Exit.Failure(cause) if cause.interrupted => ZIO.halt(cause)
      case Exit.Failure(_)                          => fb
    }
```

Reply:
It doesn't work with `onCancel` - the old problem resurfaces

Reply:
Really? All the tests passed for me locally?

I used:

```scala
  override final def forceR[A, B](fa: F[A])(fb: F[B]): F[B] =
    fa.foldCauseM(
      cause => if (cause.interrupted) ZIO.halt(cause) else fb,
      _ => fb
    )

  override final def onCancel[A](fa: F[A], fin: F[Unit]): F[A] =
    fa.onError(cause => fin.orDieWith(fiberFailure).unless(cause.failed))

  override final def guarantee[A](fa: F[A], fin: F[Unit]): F[A] =
    fa.ensuring(fin.orDieWith(fiberFailure))

  override final def bracket[A, B](acquire: F[A])(use: A => F[B])(release: A => F[Unit]): F[B] =
    acquire.bracket(release.andThen(_.orDieWith(fiberFailure)), use)
```

Reply:
Ok I will try again later - maybe they timed out because I was running something else in the background

Reply:
The timeout for the tests is extremely short at 2 seconds. By comparison in ZIO we use 60 seconds. I noticed one test `ZIO interruption is tied to F interruption` failing on CI but wasn't able to replicate it locally with `nonFlaky(100000)` and a longer timeout.